src/main.cpp

```diff
@@ -202,8 +202,20 @@ main(int argc, char *argv[])
     QCommandLineParser parser;
     parser.addHelpOption();
     parser.addVersionOption();
-    QCommandLineOption debugOption(QStringLiteral("debug"), QStringLiteral("Enable debug output"));
+    QCommandLineOption logLevel(
+      QStringLiteral("loglevel"),
+      QStringLiteral("Set the log level, or a <component>=<level> list like \"warn,ui=off\". "
+                     "levels:{trace,debug,info,warning,error,critical,off} "
+                     "components:{crypto,db,net,qml,ui}"),
+      QStringLiteral("loglevel"));
+    parser.addOption(logLevel);
+    QCommandLineOption debugOption(QStringLiteral("debug"),
+                                   QStringLiteral("Alias for \"--loglevel trace\"."));
     parser.addOption(debugOption);
+    QCommandLineOption noLog(QStringLiteral("nolog"), QStringLiteral("No log file output."));
+    parser.addOption(noLog);
+    QCommandLineOption noStderr(QStringLiteral("nostderr"), QStringLiteral("No stderr output."));
+    parser.addOption(noStderr);
```

Comment:
How about instead of making this 2 options, we make it a list, i.e.: "--log-output=file,stderr"?

Reply:
Hm...

What's your goal with that approach? 

I thought at first that it might make the --help text slightly shorter, but on further consideration, I suspect that describing the valid arguments would end up taking two lines anyway.

It would also make command lines longer.

It would also require a string tokenizer to parse the commas. (We get this for free in the loglevel argument because spdlog handles it internally.) I was hoping to keep this simple, minimizing the added code and the time I spend on it.

C's strtok() would do the job, but would come with the risks of using C-style strings in the code base. I'm not aware of a standard C++ equivalent. QStringTokenizer() doesn't exist in Qt 5. I guess QString::split() would be the way to go if I were to write a tokenizer.

How important is this to you?

Reply:
My vote is -1 on this suggestion, mainly for usability reasons.

The deciding factor for me is that it's much easier to type --nolog than --log-output=stderr every time I run nheko from the command line, which I do often when testing new code.

Writing a shell completion script wouldn't solve that, since completion would require two steps: --l<TAB>s<TAB> .

Reply:
Well, splitting a string in Qt is just str.split(','), which doesn't sound like that much effort. My issue is mostly that the parameter looks weird, so I guess it is cosmetical. Doing --log-no-stderr might be an acceptable alternative as well or something like that? I would assume you use completion anyway to just run your previous command.

Reply:
> I would assume you use completion anyway to just run your previous command.

If that were the case, I wouldn't be spending time arguing the point. ;)

Command history can somewhat gloss over an annoying command line option, but only if I'm running exactly the same command line every time. When I have to edit it or iterate through the history to find the right one, it doesn't help. (I've been doing this a lot lately while comparing palettes, for example.)

Tab completion can sometimes help with annoying command line options, but only if they have unique prefixes, and only if a completion script exists (it currently doesn't) for the shell I'm in (I don't always use bash).

> My issue is mostly that the parameter looks weird, so I guess it is cosmetical. Doing --log-no-stderr might be an acceptable alternative as well or something like that?

I see. I guess the divide here is that you're more concerned with the appearance, while I'm more concerned with ease of use. (And just to be clear, I do value good aesthetics, so I sympathize with your position. :)

I don't love --log-no-stderr, but maybe we can still find some middle ground.

How about --stderr-off and --log-off? That approach has the word separation that you like, without being terribly long. The word "off" also matches spdlog's log level name, making it easy to remember. The dashes between words would still be annoying to type, but the unique prefixes would enable tab completion in any shells for which someone writes completion scripts.

Reply:
Well, log-off is confusing. That turns of stdout or the logfile? And stderr-off just looks weird to me for something that is supposed to affect logs! I don't type different Nheko commands that often, but I need to explain stuff a hundred times over, so self explanatory stuff is pretty important to me! :D

Reply:
> Well, log-off is confusing. That turns of stdout or the logfile?
> And stderr-off just looks weird to me for something that is supposed to affect logs!

Hm... It didn't occur to me that they could confuse a user, because --stderr-off and --log-off are explicitly named after the thing they affect, and because I don't think of stderr or stdout as logs at all. (They only become logs if someone redirects them to a file, but I guess the fact that we use a logging system internally to generate their output might make it easy for programmers to think of them as logs.)

Okay, I'll think about alternatives.

Reply:
I've been thinking about ways to refine your original idea to be less burdensome to type. So, instead of `--log-output=file,stderr`, how about this:

1. Starting with a concise option name would help a lot by itself: `--log=file,stderr`
2. Accepting abbreviated values would further reduce typing overhead, so these would work as well: `--log=std` or even `--log=f,s` (The help text would show the full value names, of course.)
3. An argument starting with `/` could be treated as a custom log file location: `--log=/tmp/debug.log` (This is somewhat common already in the unix world.)
4. The option could be accepted multiple times: `--log=/tmp/debug.log --log=stderr`
5. To disable output entirely: `--log=none` or maybe `--log=off` if we want to adopt spdlog's keyword.

Of course, implementing this would be more work for me, but if it yields something that makes both of us happy, I'm willing to do it.

Reply:
I think I would leave out the abbreviated options and treating stuff with a / as a filename might be a bit too much magic, since that won't work with logfiles in cwd or on Windows. Also having 3 options to control log verbosity seems like a bit much. But I guess --log=file,stderr is fine, although I still prefer a common prefix like --log-to=file.

In the end I think it doesn't make sense to make this too featurefull. Almost nobody will be using it and I don't want this to ever break.

Reply:
Windows paths: Yes, and drive letters as well. Relative paths like those in cwd: Yep. Part of the "more work for me" would have been handling those issues, of course.

Your point about too much magic is well taken, though. Consider those ideas abandoned. :) Expect a new draft soon.